posthog/api/user.py

```diff
@@ -177,14 +184,28 @@ def update(self, instance: "User", validated_data: Any) -> Any:
             validated_data["current_team"] = current_team
             validated_data["current_organization"] = current_team.organization
 
+        require_verification_feature = (
+            posthoganalytics.feature_enabled("require-email-verification", str(instance.distinct_id)) and is_cloud()
+        )
+
         if (
             "email" in validated_data
             and validated_data["email"].lower() != instance.email.lower()
             and is_email_available()
         ):
-            send_email_change_emails.delay(
-                timezone.now().isoformat(), instance.first_name, instance.email, validated_data["email"]
-            )
+            if require_verification_feature:
+                # Verification tokens remain valid until the user logs in again. So in order to invalidate existing tokens
+                # we need to log the user in first, and then generate a new token. This prevents people
+                # from generating a token for an email they have access to, then changing the email to one they don't
+                # and using the first token to verify the second email.
+                login(self.context["request"], instance, backend="django.contrib.auth.backends.ModelBackend")
```

Comment:
This hack seems to be a result of `default_token_generator` being an instance of `PasswordResetTokenGenerator` – which, as the name suggests, is tuned for the password reset case, and isn't a perfect match for the email verification case. But most of`PasswordResetTokenGenerator`  definitely is super useful, so for our purposes here I propose we subclass it and only swap the `_make_hash_value()` method to return `f"{user.pk}{user.email}{user.pending_email}{timestamp}"` instead of `f"{user.pk}{user.password}{login_timestamp}{timestamp}{email}"`. This would remove the dependency on the login timestamp or password, which are irrelevant for email verificatio,n while adding a dependency on `pending_email`, which is crucial.

See `PasswordResetTokenGenerator` code: https://github.com/django/django/blob/5b3d3e400ab9334ba429ca360c9818c6dfc3a51b/django/contrib/auth/tokens.py#LL98C9-L98C25